tests/core.test.js

```diff
@@ -0,0 +1,183 @@
+  method2: jest.fn(),
+  method3: jest.fn(),
+};
+const mockAutocompleteFunctions = {
+  auto1: jest.fn(),
+  auto2: jest.fn(),
+  auto3: jest.fn(),
+};
+
+const mockAction = {
+  id: "mockAction",
+  params: [],
+};
+const mockSettings = { id: "mockSettings" };
+
+const mockServiceConstructor = jest.fn();
+
+const mockAutocompleteParams = [{
+  value: "param1 value",
+  name: "param1",
+  type: "string"
+}];
+const mockAutocompleteSettings = [{
+  value: "setting1 value",
+  name: "setting1",
+  valueType: "string"
+}];
+
+const mockClient = { id: "mockClient" };
+const mockParsedParams = { id: "mockParsedParams" };
+const mockRegion = "mock region";
+const mockQuery = "mock query";
+
+beforeEach(() => {
+  core = rewire("../core.js");
+})
+
+describe("mapToAwsMethod", () =>{
+  const mockMethodPromise = jest.fn();
+
+  const mockClient = {
+    mockMethod: jest.fn().mockImplementation((payload) => ({
+      promise: mockMethodPromise
+    })),
+    anotherMockMethod: jest.fn().mockImplementation(() => ({
+      promise: () => {}
+    })),
+  };
+
+  const simpleParameters = {
+    name: "some name"
+  };
+
+  const parametersWithoutEmptyValues = {
+    string: "Some text",
+    number: 5,
+    bool: true,
+    array: ["Some value", "Some other value"],
+    nestedObject: {
+      key: "value"
+    }
+  };
+
+  test("returns a function that triggers only specified service method on call with provided parameters", () => {
+
+    const returnedFunction1 = core.mapToAwsMethod("mockMethod");
+    const returnedFunction2 = core.mapToAwsMethod("anotherMockMethod");
+
+    returnedFunction1(mockClient, parametersWithoutEmptyValues);
+
+    expect(mockClient.mockMethod).toHaveBeenCalledTimes(1);
+    expect(mockClient.mockMethod).toHaveBeenCalledWith(parametersWithoutEmptyValues);
+    expect(mockClient.anotherMockMethod).toHaveBeenCalledTimes(0);
+
+    returnedFunction2(mockClient, simpleParameters);
+
+    expect(mockClient.mockMethod).toHaveBeenCalledTimes(1);
+    expect(mockClient.anotherMockMethod).toHaveBeenCalledTimes(1);
+    expect(mockClient.anotherMockMethod).toHaveBeenCalledWith(simpleParameters);
+  });
+
+  test("returns a function that triggers the payload function with params on call and pass its result to service method", () => {
+    const mockPayloadFunction = jest.fn().mockImplementation((params) => parametersWithoutEmptyValues);
+    const returnedFunction = core.mapToAwsMethod("mockMethod", mockPayloadFunction);
+    const someRegion = "some region";
+
+    expect(mockPayloadFunction).toHaveBeenCalledTimes(0);
+
+    returnedFunction(mockClient, simpleParameters, someRegion);
+    expect(mockPayloadFunction).toHaveBeenCalledTimes(1);
+    expect(mockPayloadFunction).toHaveBeenCalledWith(simpleParameters, someRegion);
+    expect(mockClient.mockMethod).toHaveBeenCalledWith(parametersWithoutEmptyValues);
+  });
+
+  test("returns a function that calls `promise()` method on service function result on call", () => {
```

Comment:
Afaik `promise` is always called when `mapToAwsMethod` is used. Can't we incorporate that into remaining functions? It's a core effect of the function we're testing.

Reply:
I don't think we can. `promise` is the method that exists on every result coming from AWS client, but we cannot assume that AWS method result is always returned. The are valid scenarios where we're simply returning an object which combines various results from various AWS calls -- see EC2 `createRouteTableWorkflow` as an example.